pkg/provider/kubernetes/gateway/client.go

```diff
@@ -0,0 +1,444 @@
+			}
+		}
+
+		for t, ok := range c.factoriesSecret[ns].WaitForCacheSync(stopCh) {
+			if !ok {
+				return nil, fmt.Errorf("timed out waiting for controller caches to sync %s in namespace %q", t.String(), ns)
+			}
+		}
+	}
+
+	return eventCh, nil
+}
+
+func (c *clientWrapper) GetHTTPRoutes(namespace string, selector labels.Selector) ([]*v1alpha1.HTTPRoute, error) {
+	if !c.isWatchedNamespace(namespace) {
+		return nil, fmt.Errorf("failed to get HTTPRoute %s with labels selector %s: namespace is not within watched namespaces", namespace, selector)
+	}
+	httpRoutes, err := c.factoriesGateway[c.lookupNamespace(namespace)].Networking().V1alpha1().HTTPRoutes().Lister().HTTPRoutes(namespace).List(selector)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(httpRoutes) == 0 {
+		return nil, fmt.Errorf("failed to get HTTPRoute %s with labels selector %s: namespace is not within watched namespaces", namespace, selector)
+	}
+
+	return httpRoutes, nil
+}
+
+func (c *clientWrapper) GetGateways() []*v1alpha1.Gateway {
+	var result []*v1alpha1.Gateway
+
+	for ns, factory := range c.factoriesGateway {
+		gateways, err := factory.Networking().V1alpha1().Gateways().Lister().List(labels.Everything())
+		if err != nil {
+			log.WithoutContext().Errorf("Failed to list Gateways in namespace %s: %v", ns, err)
+			continue
+		}
+		result = append(result, gateways...)
+	}
+
+	return result
+}
+
+func (c *clientWrapper) GetGatewayClasses() ([]*v1alpha1.GatewayClass, error) {
+	return c.factoryGatewayClass.Networking().V1alpha1().GatewayClasses().Lister().List(labels.Everything())
+}
+
+func (c *clientWrapper) UpdateGatewayClassStatus(gatewayClass *v1alpha1.GatewayClass, condition metav1.Condition) error {
+	gc := gatewayClass.DeepCopy()
+
+	var newConditions []metav1.Condition
+	for _, cond := range gc.Status.Conditions {
+		// No update for identical condition.
+		if cond.Type == condition.Type && cond.Status == condition.Status {
+			return nil
+		}
+
+		// Keep other condition types.
+		if cond.Type != condition.Type || cond.Status != condition.Status {
+			newConditions = append(newConditions, cond)
+		}
+	}
+
+	// Append the condition to update.
+	newConditions = append(newConditions, condition)
+	gc.Status.Conditions = newConditions
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	_, err := c.csGateway.NetworkingV1alpha1().GatewayClasses().UpdateStatus(ctx, gc, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to update GatewayClass %q status: %w", gatewayClass.Name, err)
+	}
+
+	return nil
+}
+
+func (c *clientWrapper) UpdateGatewayStatus(gateway *v1alpha1.Gateway, gatewayStatus v1alpha1.GatewayStatus) error {
+	if !c.isWatchedNamespace(gateway.Namespace) {
+		return fmt.Errorf("cannot update Gateway status %s/%s: namespace is not within watched namespaces", gateway.Namespace, gateway.Name)
+	}
+
+	g := gateway.DeepCopy()
+
+	if statusEquals(g.Status, gatewayStatus) {
```

Comment:
I'm wondering here if it's appropriate to return an `error` if the statuses are equal?
Maybe we just have to return?